crates/walrus-indexer/src/async_task_manager.rs

```diff
@@ -0,0 +1,705 @@
+                            tracing::error!("No more tasks available");
+                            tokio::time::sleep(inter_task_delay).await;
+                            continue;
+                        }
+                        Err(e) => {
+                            tracing::error!("Failed to get next task: {}", e);
+                            tokio::select! {
+                                _ = cancel_token.cancelled() => break,
+                                _ = tokio::time::sleep(inter_task_delay) => continue,
+                            }
+                        }
+                    };
+
+                    if cancel_token.is_cancelled() {
+                        break;
+                    }
+
+                    // Wait for a permit (this blocks until one is available).
+                    let permit = match semaphore.clone().acquire_owned().await {
+                        Ok(p) => p,
+                        Err(_) => {
+                        tracing::error!("Failed to acquire permit");
+                            break;
+                        }
+                    };
+
+                    tracing::debug!("Processing task {:?}", task);
+
+                    // Spawn task processing in background.
+                    Self::spawn_task_worker(
+                        task,
+                        permit,
+                        store.clone(),
+                        executor.clone(),
+                        Some(retry_notification_sender.clone()),
+                    );
+
+                    tokio::time::sleep(inter_task_delay).await;
+                }
+            }
+        }
+    }
+
+    /// Spawn a worker to execute a task with the given permit.
+    fn spawn_task_worker(
+        task: T,
+        permit: tokio::sync::OwnedSemaphorePermit,
+        store: Arc<S>,
+        executor: Arc<E>,
+        retry_notification_sender: Option<mpsc::UnboundedSender<()>>,
+    ) {
+        tokio::spawn(async move {
+            tracing::debug!("Worker started for task {:?}", task);
+
+            // Execute the task.
+            let result = executor.execute(task.clone()).await;
+
+            // Handle the result.
+            match result {
+                Ok(_) => {
+                    tracing::debug!("Task {:?} executed successfully", task);
+                }
+                Err(e) => {
+                    tracing::info!("Task {:?} failed: {}, adding to retry queue", task, e);
+
+                    // Move to retry queue for retry processing.
+                    if let Err(e) = store.move_to_retry_queue(&task).await {
+                        tracing::error!("Failed to move task to retry queue: {}", e);
+                    }
+
+                    // Signal that retry processing is needed.
+                    if let Some(sender) = retry_notification_sender {
+                        let _ = sender.send(());
+                    }
```

Comment:
question: I'm a bit confused by normal task and retry task now. Can they be executed at the same time? Will there be any ordering constraints?